### PR TITLE
[CARBONDATA-781] Reuse SegmentProperties object when schema is the same in table level to reduce the memory

### PR DESCRIPTION
**Pull request by watermen**
When I load carbondata 1000+ times with 35 nodes, I found SegmentProperties objects occupy 2.5+G(76K * 35 * 1000) memory in driver. 
carbonproperties
I don't have small files so I don't want to compact the segments. I analyzed the dump file and found the values of SegmentProperties are the same in same segment, so I think we can reuse the SegmentProperties object in same segment.

cc @jackylk @QiangCai

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1168/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1176/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1186/

**Comment by jackylk** [comment]
Earlier, this was used for handling restructure information. Now it is handled in another way, so yes we should change it to reduce the number of objects

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1200/

**Comment by watermen** [comment]
@jackylk You mean now we can store properties in table level(Maybe called TableProperties) insteads of SegmentProperties?

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1201/

**Comment by jackylk** [comment]
I checked recently merged PR641 for restructuring, it seems that this is still used for maintaining metadata for each Segment, so it can’t be changed to TableProperties.
One suggestion I have is to abstract and re-use some of the big object inside SegmentProperties, like the CarbonDimension and CarbonMeasure objects. They should be cached and get by ID, then it can be re-used across Segments.


发件人: Yadong Qi [mailto:notifications@github.com]
发送时间: 2017年3月17日 10:00
收件人: apache/incubator-carbondata
抄送: Likun (Jacky); Mention
主题: Re: [apache/incubator-carbondata] [CARBONDATA-781] Reuse the same SegmentProperties objects to reduce the memory (#659)


@jackylk<https://github.com/jackylk> You mean now we can store properties in table level(Maybe called TableProperties) insteads of SegmentProperties?

**Comment by watermen** [comment]
@jackylk If I want to reuse `SegmentProperties` and only check the parameters `columnsInTable` and `columnCardinality`, which case I will missing(wrong case)? Because compare `SegmentProperties` objects are easier than `CarbonDimension` and `CarbonMeasure` objects.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1334/

**Comment by watermen** [comment]
@jackylk @QiangCai I have already modified code with "Store one SegmentProperties object each segment" solution.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1335/

File: core/src/main/java/org/apache/carbondata/core/datastore/block/SegmentTaskIndex.java [hunk 1]
```diff
@@ -16,30 +16,52 @@
...
 import org.apache.carbondata.core.datastore.impl.btree.BlockBTreeBuilder;
+import org.apache.carbondata.core.metadata.AbsoluteTableIdentifier;
 import org.apache.carbondata.core.metadata.blocklet.DataFileFooter;
 
 /**
  * Class which is responsible for loading the b+ tree block. This class will
  * persist all the detail of a table segment
  */
 public class SegmentTaskIndex extends AbstractIndex {
+  private static Map<SegmentKey, SegmentProperties> segmentPropertiesCached =
```
Comment (QiangCai) [inline core/src/main/java/org/apache/carbondata/core/datastore/block/SegmentTaskIndex.java]: why not use TableSegmentUniqueIdentifier?

Comment (watermen) [inline core/src/main/java/org/apache/carbondata/core/datastore/block/SegmentTaskIndex.java]: I didn't find the class before, I will try to add a segmentproperites in TableSegmentUniqueIdentifier.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1351/

File: core/src/test/java/org/apache/carbondata/core/datastore/block/SegmentTaskIndexTest.java [hunk 2]
```diff
@@ -58,7 +58,9 @@
       @Mock public void build(BTreeBuilderInfo segmentBuilderInfos) {}
     };
     long numberOfRows = 100;
-    SegmentTaskIndex segmentTaskIndex = new SegmentTaskIndex();
+    SegmentProperties properties = new SegmentProperties(footerList.get(0).getColumnInTable(),
```
Comment (QiangCai) [inline core/src/test/java/org/apache/carbondata/core/datastore/block/SegmentTaskIndexTest.java]: should be after the initialization of variable footerList.
move to line 72

**Comment by kumarvishal09** [comment]
@watermen Attached heap dump is of executor side btree or driver side? 
1. Because in driver side btree is loaded based on segment and one segment will have only one segment property instance. 
2. In executor side carbon is loading segment property for each block(carbon data file), so number of segment properties instance will be more in executor side. 

In my opinion we need to optimise in executor side btree loading , for one segment across blocks there should be only one segment property instance

**Comment by watermen** [comment]
@kumarvishal09 
1. In the driver side, one segment has n(the number of nodes) SegmentProperties objects. You can see the `SegmentTaskIndexStore.loadBlocks` or ask @QiangCai for detail.
2. In the executor side, we can't just remain one SegmentProperties object each segment, because tasks in same segment will run on the different executor, means different progress.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1356/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1359/

**Comment by QiangCai** [comment]
@kumarvishal09 dump picture is driver tree.
@watermen this pr only implement to reuse segment properties in driver side. can you try to do it in executor side?  About the building of executor side tree, please have a look AbstractQueryExecutor.initQuery and BlockIndexStore.getAll.

**Comment by watermen** [comment]
@QiangCai When we query the big table, the memory pressure in driver side is greater than executor side. So I think we can first reuse the segment properties in driver side in this pr, and do the reuse in executor side in another pr.

**Comment by kumarvishal09** [comment]
@watermen @QiangCai 
1. In driver side segment is loaded based on taskid for each segment ...here we across task id for same segment we can create load only one segment properties ...and across segments if carbon table schema is same and cardinality of dictionary column is same we can reuse the same segment properties 
2. In executor side across task for same executor if segment properties is same we can try to reuse the same segment properties

**Comment by watermen** [comment]
@kumarvishal09 I agree with your idea. And I think we should remain a static HashMap, which key is <AbsoluteTableIdentifier, List<ColumnSchema>, int[] columnCardinality>(This is my first implement). But @QiangCai think maybe we will alter the SegmentProperties object later(such as column visibility), so if we reuse the SegmentProperties object across the segments, something will be wrong.

**Comment by kumarvishal09** [comment]
@watermen you can move all the schema related properties to some wrapper class and in wrapper class implement hash code and equals based on dimension column(including complex dimension) and measure column and and in segment index store you can have a static hashmap for storing the table to list of segment properties , this will solve alter table problem

File: core/src/main/java/org/apache/carbondata/core/datastore/SegmentTaskIndexStore.java [hunk 3]
```diff
@@ -338,14 +339,28 @@ private synchronized Object addAndGetSegmentLock(String segmentId) {
...
             tableBlockInfoList, tableIdentifier);
-    AbstractIndex segment = new SegmentTaskIndex();
+
+    if (null == tableSegmentUniqueIdentifier.getSegmentProperties()) {
+      // create a metadata details
+      // this will be useful in query handling
+      // all the data file metadata will have common segment properties we
+      // can use first one to get create the segment properties
+      SegmentProperties segmentProperties =
+          new SegmentProperties(footerList.get(0).getColumnInTable(),
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datastore/SegmentTaskIndexStore.java]: I think the hashmap approach is OK. You can keep a hashmap in this class and store the mapping from a key (calculated from schema) to the SegmentProperties object.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1726/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1729/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1734/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1738/

File: core/src/main/java/org/apache/carbondata/core/datastore/SegmentTaskIndexStore.java [hunk 4]
```diff
@@ -396,4 +419,34 @@ public TaskBucketHolder(String taskNo, String bucketNumber) {
       return result;
     }
   }
+
+  public static class SegmentPropertiesWrapper {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datastore/SegmentTaskIndexStore.java]: please add comment for this class

Comment (watermen) [inline core/src/main/java/org/apache/carbondata/core/datastore/SegmentTaskIndexStore.java]: @jackylk Added.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1759/

**Comment by watermen** [comment]
@jackylk @kumarvishal09 Can you review the code again?

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**